### PR TITLE
[pruning][core][feature] Implement convert for structured pruning

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88625
* #88436

Summary:

This PR implements `convert` in the BaseStructuredSparsifier. This function will return a model with pruned linear/conv2d layers.

Before, this functionality existed as a separate object in PrunedFuser.

The new convert function works by finding instances of specific patterns in the graph we get from
symbolically tracing the model.

Once we've found a match with a pattern, we then run that patterns custom conversion function to resize the weights appropriately. The conversion function also takes care of bias propogation.

We've implemented conversion functions for conv2d and linear patterns, in `convert_functions.py`.
There may be some gaps in the patterns compared to the previous implementation, but the conversion functionality should be the same.

Also migrated the tests from the previous PrunedFuser.

## Old Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
pruned_fuser = PrunedFuser(original, pruner)
fused_model = pruned_fuser.fuse()
```

## New Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
converted_model = pruner.convert()
```

Test Plan:
```
python test/test_ao_sparsity.py -- TestBaseStructuredSparsifierConvert
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by z-a-f — CHANGES_REQUESTED** [review]
Can we have unittests for specific functionality:
- `fuse` results in a fused model
- `convert` results in a converted model

Do we have a doc describing the `convert`? I feel like we should be careful about the FC.

Most of the functions are not expected to be public. Consider prepending `_` to the names

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 1]
```diff
@@ -56,18 +157,60 @@ def _prepare(self, *args, **kwargs) -> None:
...
+                                convert_block.append(node.target)
+                        convert_fn(*convert_block)
+
+        # remove bias hooks, since biases are propogated during module conversion.
+        for handle in self.bias_handles:
+            handle.remove()
+
+        self.traced.graph.lint()
+        self.traced.recompile()
+        return self.traced
```
[the remark this answers is not in the document]
Comment (jcaip) [reply torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:216]: I think the conditional is kind of specific to structured pruning, since it checks for a FakeStructuredSparsity parameterization. Without it though, definitely agree.  

I could move the check to within the conversion functions, but then we'd be running the conversion function for every pattern match, regardless if it's parameterized and I don't like that. 

I could also move this into utilities and pass in the conditional check as function - what do you think about that?
Something like
` def apply_conversion_functions_to_matches(is_valid, ...):`

and call it like `apply_conversion_functions_to_matches(lambda x: x is not None and parametrize.is_parametrized(x) and ..., rest of args)`

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 2]
```diff
@@ -56,18 +157,60 @@ def _prepare(self, *args, **kwargs) -> None:
             )
             self.state[config["tensor_fqn"]]["mask"] = mask
             parametrize.register_parametrization(
-                module, tensor_name, parametrization(mask), unsafe=True
+                module, tensor_name, parametrization(mask)
```
[the remark this answers is not in the document]
Comment (jcaip) [reply torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:160]: It's not the exact shape, but it's guaranteed to be broadcastable since we create the mask from the shape of the tensor [here](https://github.com/pytorch/pytorch/pull/88625/files#diff-a5d99829489397e64cc60592d2be47053f5ccf0ff70c432fa6fe8b0ba6eb0306R156)

**Comment by jcaip** [comment]
> Can we have unittests for specific functionality:
> 
> * `fuse` results in a fused model
> * `convert` results in a converted model
> 
> Do we have a doc describing the `convert`? I feel like we should be careful about the FC.
> 
> Most of the functions are not expected to be public. Consider prepending `_` to the names

@z-a-f I don't think there's a fuse anymore, fuse is now convert. I don't feel strongly that this function has to be called `convert`, happy to rename it to something else (maybe `fuse` lol) if that helps with FC. 

For the unittests - is there something specific that you're looking for? Karen's tests check that convert is applied correctly, the shapes are what we expect, that masks and parameterizations are removed, and that the output is close enough to the output of the prepared model. I can refactor the tests to make them a bit more readable but I'm not sure if there's anything else to test for functionality wise. 

It would be good to have a doc for convert, I can write something. Is there a good place for that to live?

I'll update the functions to be private as well and split the PR into (unittests + utilites)  /  (convert implementation + conversion functions)

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 3]
```diff
@@ -56,18 +165,57 @@ def _prepare(self, *args, **kwargs) -> None:
...
+        for node in self.traced.graph.nodes:
+            for pattern, convert_fn in self.patterns.items():
+                matched = apply_match(modules, pattern, node, [])
+                if matched is not None:
+                    first_module = modules.get(node.target)
+
+                    # check if first module exists and has a FakeStructuredSparsity parameterization, otherwise skip
+                    if (
+                        first_module is not None
+                        and parametrize.is_parametrized(first_module)
```
Comment (vkuzo) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:205]: should this check for `FakeStructuredSparsity`?

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 4]
```diff
@@ -1,17 +1,121 @@
...
+import torch.nn.functional as F
 from torch import nn
+from torch.fx import symbolic_trace
 from torch.nn.utils import parametrize
+from typing import Type, Set, Dict, Callable, Tuple, Any
 
 from torch.ao.pruning import BaseSparsifier
 from .parametrization import FakeStructuredSparsity, BiasHook
+from .match_utils import apply_match
+from .convert_functions import (
```
Comment (vkuzo) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:12]: in AO codebase `convert` usually means `quantization convert`.  Is there a different name we could use here just to make it obvious that this is not quantization convert?

**Comment by jcaip** [comment]
Closing since this is now split into two, see #89777 and #89776
